shared/domains/chelonia/chelonia.js

```diff
@@ -792,48 +797,113 @@ export default (sbp('sbp/selectors/register', {
-        if (has(this.removeCount, contractID)) {
-          if (this.removeCount[contractID] > 1) {
-            this.removeCount[contractID] -= 1
-          } else {
-            delete this.removeCount[contractID]
+      return sbp('chelonia/private/queueEvent', contractID, () => {
+        const rootState = sbp(this.config.stateSelector)
+        const fkContractIDs = Array.from(new Set(Object.values(rootState[contractID]?._vm?.authorizedKeys ?? {}).filter((k) => {
+          return !!(k: any).foreignKey
+        }).map((k) => {
+          try {
+            const fkUrl = new URL((k: any).foreignKey)
+            return fkUrl.pathname
+          } catch {
+            return undefined
           }
-        }
-        if (!rootState.contracts[contractID].pendingRemove) {
-          return undefined
-        }
-      }
+        }).filter(Boolean)))
 
-      if (this.removeCount[contractID] >= 1) {
-        rootState.contracts[contractID].pendingRemove = true
-        return undefined
-      }
+        sbp('chelonia/private/removeImmediately', contractID)
 
-      return sbp('chelonia/private/queueEvent', contractID, [
-        'chelonia/private/removeImmediately', contractID
-      ])
+        if (fkContractIDs.length) {
+          // Attempt to release all contracts that are being monitored for
+          // foreign keys
+          sbp('chelonia/contract/release', fkContractIDs, { try: true }).catch((e) => {
+            console.error('[chelonia] Error attempting to release foreign key contracts', e)
+          })
+        }
+      })
     }))
   },
+  'chelonia/contract/retain': function (contractIDs: string | string[], params?: { ephemeral?: boolean}): Promise<*> {
+    const listOfIds = typeof contractIDs === 'string' ? [contractIDs] : contractIDs
+    if (listOfIds.length === 0) return Promise.resolve()
+    if (!params?.ephemeral) {
+      const rootState = sbp(this.config.stateSelector)
+      listOfIds.forEach((id) => {
+        if (!has(rootState.contracts, id)) {
+          this.config.reactiveSet(rootState.contracts, id, Object.create(null))
+        }
+        this.config.reactiveSet(rootState.contracts[id], 'references', (rootState.contracts[id].references ?? 0) + 1)
+      })
+    } else {
+      listOfIds.forEach((id) => {
+        if (!has(this.ephemeralReferenceCount, id)) {
+          this.ephemeralReferenceCount[id] = 1
+        } else {
+          this.ephemeralReferenceCount[id] = this.ephemeralReferenceCount[id] + 1
+        }
+      })
+    }
+    return sbp('chelonia/contract/sync', listOfIds)
+  },
+  // the `try` parameter does not affect (ephemeral or persistent) reference
+  // counts, but rather removes a contract if the reference count is zero
+  // and the contract isn't being monitored for foreign keys. This parameter
+  // is meant mostly for internal chelonia use, so that removing or releasing
+  // a contract can also remove other contracts that this first contract
+  // was monitoring.
```

Comment:
Could you further elaborate how it is possible for a contract to find itself in the state where the retain count is 0 and yet it hasn't been removed?

Is it for contracts that are sync'd without calling `retain` on them? (If such a thing is possible, note that this isn't how Objective-C works, where any time an object is created, it begins with a retain count of `1`). Perhaps some thought should be given to the possibility of setting retain counts to `1` automatically when a contract is first sync'd?

Reply:
> Could you further elaborate how it is possible for a contract to find itself in the state where the retain count is 0 and yet it hasn't been removed?

As it's mentioned in the proposal, there are _three_ separate counts. One of them is internal to Chelonia, namely the when contracts are being monitored for foreign keys. Because this counter isn't really stored as an integer (it's calculated dynamically), and it's managed by Chelonia, you can think of `try: true` as maybe decrementing that count and then removing the contract.

As you point out, the second reason why the count could be 0 (excluding foreign keys) is if `sync` is called directly. While maybe there are legitimate reasons for doing this if one looks hard enough, contract developers should _not_ call `sync` on contracts that haven't been `retain`ed.

> Perhaps some thought should be given to the possibility of setting retain counts to 1 automatically when a contract is first sync'd?

I think this could be confusing and tricky to implement.

Confusing because you should pair all your calls to `retain` with `release`. If sync magically changes the count value, it could be confusing.

On the other hand, `sync` on a non-`retain`ed contract _could_ throw an error, which is probably the correct thing to do. This is tricky because Chelonia does this when subscribing to a contract to monitor foreign keys.